crud/src/main/java/com/redhat/lightblue/crud/AbstractBulkJsonObject.java

```diff
@@ -0,0 +1,158 @@
+ */
+package com.redhat.lightblue.crud;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Collections;
+import java.util.Comparator;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+
+import com.redhat.lightblue.util.JsonObject;
+import com.redhat.lightblue.Request;
+
+/**
+ * Base class for bulk request and responses. This class contains the
+ * code common to both, as bulk request and response are structurally
+ * similar. It deals with a JSON of the form:
+ *
+ *
+ * <pre>
+ *   {
+ *     "<entries>": [
+ *         {
+ *             "seq":0,
+ *             "op": "FIND",
+ *             "<entry>": { item }
+ *         }
+ *     ]
+ *   }
+ * </pre> 
+ */
+abstract class AbstractBulkJsonObject<T extends JsonObject> extends JsonObject {
+
+    private final List<T> entries=new ArrayList<T>();
+
+    /**
+     * Returns all entries in the bulk object
+     */
+    public List<T> getEntries() {
+        return entries;
+    }
+
+    /**
+     * Sets the entries in the bulk object. Copies the collection
+     */
+    public void setEntries(List<T> x) {
+        entries.clear();
+        entries.addAll(x);
+    }
+
+    public void add(T x) {
+        entries.add(x);
+    }
+
+    /**
+     * Overriding this method allows the concrete class to add any necessary items to the entry node.
+     *
+     * @param node The object node, an element of the entries array
+     * @param entry The entry
+     */
+    protected abstract void toJsonEntryNode(ObjectNode node,T entry);
+
+    /**
+     * Returns a JSON representation of this
+     */
+    public JsonNode toJson(String entriesName, String entryName) {
+        JsonNodeFactory factory=getFactory();
+        ObjectNode node = factory.objectNode();
+        ArrayNode arr=factory.arrayNode();
+        int seq=0;
+        for(T x:entries) {
+            ObjectNode entryNode=factory.objectNode();
+            entryNode.set("seq",factory.numberNode(seq++));
+            toJsonEntryNode(entryNode,x);
+            entryNode.set(entryName,x.toJson());
+            arr.add(entryNode);
+        }
+        node.put(entriesName, arr);
+        return node;
+    }
+
+    private static class SReq {
+        private final int seq;
+        private final JsonObject entry;
+
+        public SReq(int seq,JsonObject entry) {
+            this.seq=seq;
+            this.entry=entry;
+        }
+    }
+
+    /**
+     * Parses the actual entry from the given node
+     *
+     * @param entry The entry object, an element of the entries array
+     */
+    protected abstract T parseEntry(ObjectNode entry);
+    
+    /**
+     * Parses the bulk object from the given array node
+     */
+    protected void parse(ArrayNode entriesArray) {
```

Comment:
I can't think of a reason why BulkResponse would be used as an input or vice-versa.